File: src/main/java/eu/mshade/enderman/metadata/EndermanEntityMetadataManager.java
package eu.mshade.enderman.metadata;

import eu.mshade.enderframe.entity.EntityType;
import eu.mshade.enderframe.item.ItemStack;
import eu.mshade.enderframe.metadata.entity.EntityMetadataManager;
import eu.mshade.enderframe.metadata.MetadataType;
import eu.mshade.enderman.metadata.entity.ArmorStandMetadataBucket;
import eu.mshade.enderman.metadata.entity.PlayerMetadataBucket;
import eu.mshade.enderman.metadata.entity.ZombieMetadataBucket;

public class EndermanEntityMetadataManager extends EntityMetadataManager {

    public EndermanEntityMetadataManager() {
        this.registerMetadataIndex(0, MetadataType.BYTE, (protocolBuffer, metadata) -> protocolBuffer.writeByte((byte) metadata.get()));
        this.registerMetadataIndex(1, MetadataType.SHORT, (protocolBuffer, metadata) -> protocolBuffer.writeShort((short) metadata.get()));
        this.registerMetadataIndex(2, MetadataType.INTEGER, (protocolBuffer, metadata) -> protocolBuffer.writeInt((int) metadata.get()));
        this.registerMetadataIndex(3, MetadataType.FLOAT, (protocolBuffer, metadata) -> protocolBuffer.writeFloat((float) metadata.get()));
        this.registerMetadataIndex(4, MetadataType.STRING, (protocolBuffer, metadata) -> protocolBuffer.writeString((String) metadata.get()));
        this.registerMetadataIndex(5, MetadataType.SLOT, (protocolBuffer, metadata) -> protocolBuffer.writeItemStack((ItemStack) metadata.get()));
        this.registerMetadataIndex(6, MetadataType.BLOCK_POSITION, new BlockPositionMetadataTypeBuffer());
        this.registerMetadataIndex(7, MetadataType.ROTATION, new RotationMetadataTypeBuffer());

        this.registerEntityMetadataBucket(EntityType.PLAYER, new PlayerMetadataBucket());
        this.registerEntityMetadataBucket(EntityType.ARMOR_STAND, new ArmorStandMetadataBucket());
        this.registerEntityMetadataBucket(EntityType.ZOMBIE, new ZombieMetadataBucket());

        /*
        registerEntityBuffer(EntityType.PLAYER, new PlayerMetadataBuffer());
        registerEntityBuffer(EntityType.ZOMBIE, new ZombieMetadataBuffer());
        registerEntityBuffer(EntityType.END_CRYSTAL, new EnderCrystalMetadataBuffer());
        registerEntityBuffer(EntityType.BOAT, new BoatMetadataBuffer());
        registerEntityBuffer(EntityType.SPIDER, new SpiderMetadataBuffer());

         */


        /*
        registerMetadataRewriterBucket(EntityType.ARMOR_STAND, new ArmorStandMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.ARROW, new ArrowMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.BAT, new BatMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.BLAZE, new BlazeMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.BOAT, new BoatMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.CREEPER, new CreeperMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.END_CRYSTAL, new EnderCrystalMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.ENDERMAN, new EndermanMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.FIREWORK_ROCKET, new FireworkMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.FURNACE_MINECART, new FurnaceMinecartMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.GHAST, new GhastMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.GUARDIAN, new GuardianMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.HORSE, new HorseMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.IRON_GOLEM, new IronGolemMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.ITEM_FRAME, new ItemFrameMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.ITEM, new ItemMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.MINECART, new MinecartMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.OCELOT, new OcelotMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.PIG, new PigMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.PLAYER, new PlayerMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.RABBIT, new RabbitMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.SHEEP, new SheepMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.SKELETON, new SkeletonMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.SLIME, new SlimeMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.SPIDER, new SpiderMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.VILLAGER, new VillagerMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.WITCH, new WitchMetadataRewriterBucket());
        //TODO RE-WORK THIS ! pas bon
        //duplication de code car c'est exactement la même chose pour les 3
        registerMetadataRewriterBucket(EntityType.WITHER, new WitherMetadataRewriterBucket(0));
        registerMetadataRewriterBucket(EntityType.WOLF, new WolfMetadataRewriterBucket());
        registerMetadataRewriterBucket(EntityType.ZOMBIE, new ZombieMetadataRewriterBucket());

         */
    }
}

File: src/main/java/eu/mshade/enderman/metadata/EndermanEntityMetadataManager.kt
package eu.mshade.enderman.metadata

import eu.mshade.enderframe.entity.EntityType
import eu.mshade.enderframe.item.ItemStack
import eu.mshade.enderframe.metadata.Metadata
import eu.mshade.enderframe.metadata.MetadataType
import eu.mshade.enderframe.metadata.entity.EntityMetadataManager
import eu.mshade.enderframe.world.Rotation
import eu.mshade.enderframe.world.Vector
import eu.mshade.enderman.metadata.entity.*

class EndermanEntityMetadataManager: EntityMetadataManager() {

    init {
        registerMetadataIndex(0, MetadataType.BYTE) { minecraftByteBuf, metadata -> minecraftByteBuf.writeByte((metadata.get() as Byte).toInt()) }
        registerMetadataIndex(1, MetadataType.SHORT) { minecraftByteBuf, metadata -> minecraftByteBuf.writeShort((metadata.get() as Short).toInt()) }
        registerMetadataIndex(2, MetadataType.INTEGER) { minecraftByteBuf, metadata -> minecraftByteBuf.writeInt(metadata.get() as Int) }
        registerMetadataIndex(3, MetadataType.FLOAT) { minecraftByteBuf, metadata -> minecraftByteBuf.writeFloat(metadata.get() as Float) }
        registerMetadataIndex(4, MetadataType.STRING) { minecraftByteBuf, metadata -> minecraftByteBuf.writeString(metadata.get() as String) }
        registerMetadataIndex(5, MetadataType.SLOT) { minecraftByteBuf, metadata -> minecraftByteBuf.writeItemStack(metadata.get() as ItemStack) }

        registerMetadataIndex(6, MetadataType.BLOCK_POSITION) { minecraftByteBuf, metadata ->
            val vector = metadata.get() as Vector
            minecraftByteBuf.writeInt(vector.blockX)
            minecraftByteBuf.writeInt(vector.blockY)
            minecraftByteBuf.writeInt(vector.blockZ)
        }

        registerMetadataIndex(7, MetadataType.ROTATION) { minecraftByteBuf, metadata ->
            val rotation = metadata.get() as Rotation
            minecraftByteBuf.writeFloat(rotation.yaw)
            minecraftByteBuf.writeFloat(rotation.pitch)
            minecraftByteBuf.writeFloat(rotation.roll)
        }

        registerEntityMetadataBucket(EntityType.PLAYER, PlayerMetadataBucket())

        //Hostile
        registerEntityMetadataBucket(EntityType.BLAZE, BlazeMetadataBucket())
        registerEntityMetadataBucket(EntityType.CREEPER, CreeperMetadataBucket())
        registerEntityMetadataBucket(EntityType.GHAST, GhastMetadataBucket())
        registerEntityMetadataBucket(EntityType.GUARDIAN, GuardianMetadataBucket())
        registerEntityMetadataBucket(EntityType.MAGMA_CUBE, MagmaCubeMetadataBucket())
        registerEntityMetadataBucket(EntityType.SKELETON, SkeletonMetadataBucket())
        registerEntityMetadataBucket(EntityType.SLIME, SlimeMetadataBucket())
        registerEntityMetadataBucket(EntityType.ZOMBIE, ZombieMetadataBucket())
        registerEntityMetadataBucket(EntityType.WITCH, WitchMetadataBucket())

        //Passive
        registerEntityMetadataBucket(EntityType.BAT, BatMetadataBucket())
        registerEntityMetadataBucket(EntityType.CHICKEN, ChickenMetadataBucket())
        registerEntityMetadataBucket(EntityType.HORSE, HorseMetadataBucket())
        registerEntityMetadataBucket(EntityType.OCELOT, OcelotMetadataBucket())
        registerEntityMetadataBucket(EntityType.PIG, PigMetadataBucket())
        registerEntityMetadataBucket(EntityType.RABBIT, RabbitMetadataBucket())
        registerEntityMetadataBucket(EntityType.SHEEP, SheepMetadataBucket())
        registerEntityMetadataBucket(EntityType.VILLAGER, VillagerMetadataBucket())

        //Neutral
        registerEntityMetadataBucket(EntityType.CAVE_SPIDER, CaveSpiderMetadataBucket())
        registerEntityMetadataBucket(EntityType.ENDERMAN, EndermanMetadataBucket())
        registerEntityMetadataBucket(EntityType.CAVE_SPIDER, IronGolemMetadataBucket())
        registerEntityMetadataBucket(EntityType.SPIDER, SpiderMetadataBucket())
        registerEntityMetadataBucket(EntityType.WOLF, WolfMetadataBucket())

        //Boss
        registerEntityMetadataBucket(EntityType.WITHER, WitherMetadataBucket())

        //TileEntity
        registerEntityMetadataBucket(EntityType.ARMOR_STAND, ArmorStandMetadataBucket())
        registerEntityMetadataBucket(EntityType.BOAT, BoatMetadataBucket())
        registerEntityMetadataBucket(EntityType.MINECART, MinecartMetadataBucket())
        registerEntityMetadataBucket(EntityType.MINECART_FURNACE, FurnaceMinecartMetadataBucket())
        registerEntityMetadataBucket(EntityType.DROPPED_ITEM, ItemMetadataBucket())
        registerEntityMetadataBucket(EntityType.ARROW, ArrowMetadataBucket())
        registerEntityMetadataBucket(EntityType.FIREWORK, FireworkMetadataBucket())
        registerEntityMetadataBucket(EntityType.ITEM_FRAME, ItemFrameMetadataBucket())
        registerEntityMetadataBucket(EntityType.ENDER_CRYSTAL, EnderCrystalMetadataBucket())
    }
}